parlai/core/metrics.py

```diff
@@ -385,175 +414,101 @@ def __repr__(self):
+        # (other metrics such as p@k and r@k take
+        # the value of cnt into account.)
+        for k in self.eval_pr:
+            self._add(f'hits@{k}', AverageMetric(cnts[k] > 0))
 
-    def update(self, observation, labels):
+    def _add(self, key: str, value: Optional[Metric]) -> None:
         """
-        Update metrics based on an observation and true labels.
+        Add the metric to our records.
         """
-        with self._lock():
-            self.metrics['cnt'] += 1
+        if self._hogwild:
+            self._queue.put((key, value))
+        else:
+            self.metrics[key] = self.metrics.get(key) + value
 
-        # Exact match metric.
-        correct = 0
+    def update(self, observation: Message, labels: List[str]) -> None:
+        """
+        Update metrics based on an observation and true labels.
+        """
         prediction = observation.get('text', None)
+
+        self._add('exs', SumMetric(1))
+
         if prediction is not None:
-            if _exact_match(prediction, labels):
-                correct = 1
-            with self._lock():
-                self.flags['print_prediction_metrics'] = True
-                self.metrics['correct'] += correct
-                self.metrics['correct_cnt'] += 1
-
-            # F1 and BLEU metrics.
-            if 'f1' in self.metrics_list:
-                f1 = _f1_score(prediction, labels)
-            bleu_scores = {}
-            rouge1 = rouge2 = rougeL = None
-            if 'bleu-4' in self.metrics_list:
-                bleu_scores['bleu-4'] = _bleu(prediction, labels)
-            if 'bleu-1' in self.metrics_list:
-                for i in range(3):
-                    weights = [1 / (i + 1) for _ in range(i + 1)]
-                    bleu_scores[f'bleu-{i + 1}'] = _bleu(prediction, labels, weights)
-            if 'rouge-L' in self._print_metrics_list:
-                rouge1, rouge2, rougeL = _rouge(prediction, labels)
-            with self._lock():
-                if 'f1' in self.metrics:
-                    self.metrics['f1'] += f1
-                    self.metrics['f1_cnt'] += 1
-                if 'bleu-4' in self.metrics:
-                    self.metrics['bleu-4'] += bleu_scores.pop('bleu-4')
-                    self.metrics['bleu-4_cnt'] += 1
-                if 'bleu-1' in self.metrics:
-                    for b, b_score in bleu_scores.items():
-                        self.metrics[b] += b_score
-                        self.metrics[f'{b}_cnt'] += 1
-                if 'rouge-L' in self.metrics and rouge1 is not None:
-                    self.metrics['rouge-1'] += rouge1
-                    self.metrics['rouge-1_cnt'] += 1
-                    self.metrics['rouge-2'] += rouge2
-                    self.metrics['rouge-2_cnt'] += 1
-                    self.metrics['rouge-L'] += rougeL
-                    self.metrics['rouge-L_cnt'] += 1
+            self._add('accuracy', ExactMatchMetric.compute(prediction, labels))
```

Comment:
is it worth adding .compute() as an abstract method to Metric to detail when it's needed?

Reply:
I don't think so. This is very much an idiom for these standard metrics and I don't expect it to go elsewhere.